File: Contact_Density.py
#©Group Phoenix 
print ("Enter a protein conformation file name:(eg. conformation.pdb)")
filename = input()
pdb=open(filename,"r")
file = open("newfile1.txt", "w")
for line in pdb:
    if line[:4] == 'ATOM' and line[12:16] == " CA ":
        file.write(line[30:55])
        file.write("\n")

file.close()
count=0
x = []
y = []
z = []


with open('newfile1.txt') as f:
    for line in f:
        data = line.split()
        x.append(float(data[0]))
        y.append(float(data[1]))
        z.append(float(data[2]))
        count=count+1

arr1 = []
g=0
i=0

for i in range(0,count):
    for j in range(i+2,count):
       if ((x[i] - x[j])*(x[i] - x[j])+(y[i] - y[j])*(y[i] - y[j])+(z[i] - z[j])*(z[i] - z[j]))<36:
           g=g+1
           arr1.append(1)
       else:
           arr1.append(0)
print ("Enter the native protein conformation file name(eg. native.pdb):")
filename = input()
pdb=open(filename,"r")
file = open("newfile1.txt", "w")
for line in pdb:
    if line[:4] == 'ATOM' and line[12:16] == " CA ":
        file.write(line[30:55])
        file.write("\n")

file.close()
count=0
x = []
y = []
z = []


with open('newfile1.txt') as f:
    for line in f:
        data = line.split()
        x.append(float(data[0]))
        y.append(float(data[1]))
        z.append(float(data[2]))
        count=count+1
#print(count)
arr = []
m=0
i=0

for i in range(0,count):
    for j in range(i+2,count):
       if ((x[i] - x[j])*(x[i] - x[j])+(y[i] - y[j])*(y[i] - y[j])+(z[i] - z[j])*(z[i] - z[j]))<36:
           m=m+1
           arr.append(1)
       else:
           arr.append(0)
    

ng=0
print("Contacts in given protein conformation=",g)
print("Contacts in native protein conformation=",m)


len=len(arr)
for i in range(0,len):
    if(arr[i] & arr1[i]):
        ng=ng+1
print("Common contacts=",ng)

cd=(ng/m)*100
print("Contact density % =",cd)
pdb.close
import os
os.system("pause")

File: Contact_Density.cpp
# include <iostream>
#include <fstream>
#include <string.h>
using namespace std;
int main()
{
	char file_name[25],file_nam[25];
	printf("Enter the protein conformation file(pdb file).\n");
    cin>>file_name;
	ifstream inputfile(file_name);
	string s; double x[100],y[100],z[100]; int i=0;
	while(!inputfile.eof()) 
	{
		inputfile>>s;
		if(s=="ATOM")
		{
			inputfile>>s;
			inputfile>>s;
			if(s=="CA")
			{inputfile>>s;inputfile>>s;inputfile>>s;
			inputfile>>x[i]>>y[i]>>z[i];
			i++;
	
			}
		}
		
	}
	inputfile.close();
	printf("Enter the native protein conformation file(pdb file).\n");
    cin>>file_nam;
    ifstream inputfile1(file_nam);
    double x1[100],y1[100],z1[100]; int j=0;
	while(!inputfile1.eof()) 
	{
		inputfile1>>s;
		if(s=="ATOM")
		{
			inputfile1>>s;
			inputfile1>>s;
			if(s=="CA")
			{inputfile1>>s;inputfile1>>s;inputfile1>>s;
			inputfile1>>x1[j]>>y1[j]>>z1[j];
			j++;	
			}
		}
		
	}
	inputfile1.close();
	int cn=i,nn=j,counter1[1000],counter2[1000],k=0; double count1=0,count2=0,com=0;
	for(i=0;i<cn-2;i++)
	for(j=i+2;j<cn;j++)
	{
		if(((x[i] - x[j])*(x[i] - x[j])+(y[i] - y[j])*(y[i] - y[j])+(z[i] - z[j])*(z[i] - z[j]))<36)
	 	{
			count1++;
			counter1[k]=1;
			k++;
	 	}
		else
			{
				counter1[k]=0;
				k++;
			}
	}
	k=0;
	for(i=0;i<nn-2;i++)
	for(j=i+2;j<nn;j++)
	{
		if(((x1[i] - x1[j])*(x1[i] - x1[j])+(y1[i] - y1[j])*(y1[i] - y1[j])+(z1[i] - z1[j])*(z1[i] - z1[j]))<36)
	 	{
			count2++;
			counter2[k]=1;
			k++;
	 	}
		else
			{
				counter2[k]=0;
				k++;
			}
	}
 	cout<<"No. of contacts in given protein conformation="<<count1<<endl;
 	cout<<"No. of contacts in native protein conformation="<<count2<<endl;
 	for(i=0;i<k;i++)
 		if(counter1[i] && counter2[i])
 			com++;
 	cout<<"Common contacts in conformation and native protein="<<com<<endl;
 	cout<<"Contact Density="<<(com/count2)*100<<endl;
	system("pause");	
	
return 0;
}
